docs/RFCS/20220426_rbac_extensions_system_privileges.md

```diff
@@ -0,0 +1,137 @@
+
+In short, role options should be reserved for user administration permissions, anything that reads/writes from the database should be a grant.
+
+Role Option | Inheritable
+---- | ----
+CREATEROLE / NOCREATEROLE | Unchanged
+PASSWORD | Unchanged
+LOGIN / NOLOGIN | Unchanged
+VALIDUNTIL | Unchanged
+CONTROLJOB / NOCONTROLJOB | Replace with per job grants Example: GRANT PAUSE ON JOB TYPE BACKUP
+CONTROLCHANGEFEED / NOCONTROLCHANGEFEED | Changefeeds are created per table and write to a sink. We can support CHANGEFEED privilege on tables or on the database and also introduce grants on sinks
+CREATEDB / NOCREATEDB | Unchanged (Is a Postgres roleoption)
+CREATELOGIN / NOCREATELOGIN | Unchanged (?)
+VIEWACTIVITY / NOVIEWACTIVITY | Replace with system level grant
+CANCELQUERY / NOCANCELQUERY | Replace with system level grant
+MODIFYCLUSTERSETTING / NOMODIFYCLUSTERSETTING | Replace with system level grant
+VIEWACTIVITYREDACTED / NOVIEWACTIVITYREDACTED | Replace with system level grant
+SQLLOGIN / NOSQLLOGIN | Unchanged
+VIEWCLUSTERSETTING / NOVIEWCLUSTERSETTING | Replace with system level grant 
+
+##  New System Table
+
+We will create a new “system.privileges” table.
+ `
+CREATE TABLE system.privileges (
+username STRING NOT NULL,
+object INT NOT NULL,
+object_type system.object_type NOT NULL,
+privileges VARBIT NOT NULL,
+role_options VARBIT NOT NULL,
+CONSTRAINT "primary" PRIMARY KEY (username, object, object_type)
+);`
+
+`system.object_type` will be an enum that defines the various object types that can support privileges in CockroachDB. This tentatively looks like 
+```
+CREATE TYPE system.object_type AS ENUM('system', 'jobs', 'changefeed_sink')
+```
+
+Note that this will be the first example of a system enum.
+
+
+The system.privileges table will store privileges for non-descriptor backed objects.
+We will require object_type + id to uniquely identify them. The privileges will be represented as a bit mask. We can serialize and deserialize synthetic privilege descriptors using rows in this table to leverage our existing code that manipulates PrivilegeDescriptors for grants. We can also consider making the privileges and role_options columns have type `[]STRING`, the downside here is that it takes slightly more space (probably not a concern) and that we have to add some logic to serialize the string array into a bit array which is how we currently manipulate privileges. 
+
+### Alternate considerations
+The system level grants could be stored on a PrivilegeDescriptor on the system database, however this would require us to support mutations on the system database descriptor which we currently do not do and this allows us to not perform a round trip whenever we look for a system table.
+
+We could create descriptors to represent every object we want to grant on. This is not an attractive option at first glance as these descriptors will likely only be used to store privileges and nothing more.
```

Comment:
does it make sense to store privileges with the object_type, e.g. for jobs, add an extra column on `system.jobs` with a privilege protobuf? that way everything is tied together, but i guess the logic can get more confusing.

Reply:
I guess an alternative is to make a new system table for each object type?
I don't think `system.jobs` makes sense to add privileges to since it's not exactly per job type.

The downside of a new system table per object type is that if another team wants to add a new object type, instead of adding to an enum they'd have to create a whole new table.